TODO/getting-to-know-css-grid-layout.md

```diff
@@ -1 +1,312 @@
+
+```
+.grid {
+    display: grid;
+    grid-template-columns: 1fr 1fr;
+}
+```
+
+为了让子元素从右侧开始，我们设置子元素从第2列开始：
+
+```
+.child {
+    grid-column-start: 2;
+}
+```
+
+**注意**：在Firefox 52中的一个差异导致一个垂直对齐问题——[基于FR单位的行不会拉伸得跟整个窗口一样高](https://bugzilla.mozilla.org/show_bug.cgi?id=1346699)。为了解决（address）这个问题我们设置子元素为栅格项，并给每一行设置一个想要的高度：
+
+```
+.l-grid--full-height {
+    grid-template-rows: minmax(100vh, 1fr);
+}
+```
+
+[设置留白示例](http://codepen.io/chriswrightdesign/pen/5a9bc28c370ad5155515c3c7a19653d8)
+
+## 用内容死区（content dead-zones）创建空白
+在四列布局中，给本来在第三列的栅格项上设置`grid-column-start:2;`，那么会找到下一个可用的第二列来填充空间。
+
+栅格轨道会跳过某些列，直到找到下一列。我们可以利用这个方法在栅格内创建空白，没有内容的栅格也会被分配。
+[\[创建空白示例\]](http://codepen.io/chriswrightdesign/pen/427c3e51c320e04662be56dad44dee74)
+
+![](http://n1image.hjfile.cn/res7/2017/04/20/90fcea0b53d4712f827c74ec950cebd1.jpg)
+[使用grid-template-columns 和 grid-column-start创建空白]
+
+## 创建行
+如果我们想分割布局为四份，我们目前所了解的关于列的布局方式对行同样有用：
+
+```
+.grid {
+    display: grid;
+    grid-template-columns: 1fr 1fr;
+    grid-template-rows: 250px 250px;
+}
+```
+
+![](http://n1image.hjfile.cn/res7/2017/04/20/1e6446cad82bb945bb7f559fa0f869e8.jpg)
+[同时使用grid-template-columns 和 grid-template-rows创建栅格布局]
+
+理想情况下这个示例是没问题的。因为此时每个栅格项的内容足够少而不会撑开每行。但随着内容的变化，一切都不一样了。当示例中的内容超出指定行的大小后，看下会发生什么：
+
+![](http://n1image.hjfile.cn/res7/2017/04/20/a2bce80e7d94d2b9f2931e061dff7780.jpg)
+[内容超出声明的行高]
+
+我们创建了250px高的两行，如果内容超过每行的高度，将会打破布局并和后面的行的内容重叠。并不是一个我们想要的结果。
+
+## 灵活的设置最小值
+这里的场景是设置了最小值，但是允许我们根据内容弹性变化。这里我们通过上面旧浏览器示例中的`minmax`关键字实现。
+
+```
+.grid {
+    grid-template-rows: minmax(250px, auto) minmax(250px, auto);
+}
+```
+[创建有最小值的弹性行](http://codepen.io/chriswrightdesign/pen/447e07c84aac9abb5514f9b9fbb18459)
+
+
+现在我们已经了解了创建带有内容的行的基础方法，我们开始来创建水平和垂直交错的更复杂栅格布局。
+
+![](http://n1image.hjfile.cn/res7/2017/04/20/cab8e4f4681e3841bcd7865650240058.jpg)
+\[使用grid-column-start和span关键字创建复杂栅格布局[Unsplash](https://unsplash.com/)]
+
+## 创建更复杂的栅格
+我们开始创建更复杂的栅格布局。将栅格中的每个栅格项设置成占据多条轨道，在一列内，我们能通过`grid-column-start`和`grid-column-end`实现，或者通过如下所示更简单的写法：
+
+```
+grid-column: 1 / 3;
+```
+
+用这种实现方式的弊端是并不是很“模块化”，为了定位每块内容需要写很多代码。`span`关键字更符合模块化的思路，因为我们能放在任何地方，让栅格来控制他。我们可以定义栅格项的开始位置，及其占据的轨道数：
+
+```
+.span-column-3 {
+    grid-column-start: span 3;
+}
+```
+
+任何添加该class的栅格将会从其开始位置，占据三个轨道。
+
+[\[通过span实现的复杂栅格\]](http://codepen.io/chriswrightdesign/pen/35102dd4a2cf7fbadb33dd53cb88787d)
+
+## 使用span设计一个布局
+我们能设计一个多轨道布局，通过将栅格布局打散成最小单位。本示例中的最小单位是图中高亮的部分。
+
+![](http://n1image.hjfile.cn/res7/2017/04/20/028fe538b6c135f129441ad7f7a0f3a6.jpg)
+[通过最小栅格单位结合span创建更大的栅格]
+
+围绕最小单位，我们能灵活的使用span来创建一些有意思的布局，因为span是可以叠加的——你可以结合列和行的轨道在栅格中创建多层级。
+
+## 不需要媒体查询（media queries）的弹性栅格
+虽然上面说到的例子能在可用空间内适应变化，但是没有一个是专门为空间变化设计的。栅格有两个非常有用的特性来适应可用空间的变化。这两个属性叫`auto-fit`和`auto-fill`，像下面这样结合`repeat function`和`minmax function`使用：
+
+```
+grid-template-columns: repeat(auto-fit, minmax(250px, 1fr));
+```
+
+这些值代替了repeat中的数字，并计算在每条轨道上会填充多少行或列。二者之间最大不同是当一条轨道上空白的溢出时的他们的处理方式不同。
+
+`auto-fit`尝试放最大的数量的重复元素待一列内且该列能自动适配而不溢出。当没有足够的空间来放置更多的元素时，之后的元素将会放到下一行，不能填满的空间将会空着。
```

Comment:
Auto-fill attempts to put the largest number of repeated items into a column that it can manage without overflowing. 
->
Auto-fill尝试在不导致列溢出的情况下,放置该列能处理的最大数量的重复元素.